Feature: Literal ops and conversions

liss-h opened the pull request.

Clueliss committed:
Groundwork for Literal operations
- Added subtypes LiteralDatatype that require numeric or logical operations
- added automatic registration of those additional capabilities
- changed definitions of xsd types to reflect their capabilities
- added tests for those capabilities

Clueliss committed:
added tests for numeric ops and type conversions

Clueliss committed:
added functions to enable the specification of node storage for operators

Clueliss committed:
implemented ops in literal

Clueliss committed:
needed to fixed equality operator, or else the hierarchy tests will always pass
added missing capability in test

Clueliss committed:
fixed a bug where never explicitly naming a type that will be used during a promotion does not instantiate it's template, and thus creates error during promotion

Clueliss committed:
added some final comments, and minor improvements

Clueliss committed:
added legend to hierarchy test

Clueliss committed:
added additional asserts and other small improvements

Clueliss committed:
added noexcept guarantees and renamed tparams

Clueliss committed:
added utilities that will be needed for type conversions

Clueliss committed:
implemented compiletime conversion table generation

Clueliss committed:
modified registry to use compiletime conversion table to generate it's runtime representation

Clueliss committed:
modified high level interface to use the new registry conversion functionality

Clueliss committed:
disabled gcc10 for tests, because the compiler segfaults

Clueliss committed:
simplified conversion searching in DatatypeRegistry

Clueliss committed:
removed unnecessary items in Util and added type trait to check for equality of multiple types

Clueliss committed:
optimizations, more comments and concepts for compiletime conversion table generation

Clueliss committed:
fixes in Literal to use simplified registry correctly

Clueliss committed:
removed 1 constexpr (which wasn't necessary) to make gcc11 happy because it does not have constexpr string ==

Clueliss committed:
Merge branch 'develop' into feature/literal-ops-comptime

Clueliss committed:
now returning null Literal instead of throwing exceptions in Literal operations

Clueliss committed:
fixed a bug where error values (null literals) would not be carried through operations and instead triggered an assertion

Clueliss committed:
had to remove using decl for enum variants because older compilers don't support it.

bigerl commented on src/rdf4cpp/rdf/datatypes/registry/Util.hpp:
```diff
@@ -0,0 +1,112 @@
+ * @return the tuple generated by applying f to each element in tup
+ */
+template<typename F, typename TupleLike>
+constexpr auto tuple_map(TupleLike &&tup, F &&f) {
```
A concept for tuple would make compiler errors more readable.

bigerl commented on src/rdf4cpp/rdf/datatypes/registry/Util.hpp:
```diff
@@ -0,0 +1,112 @@
+ * @param f a binary operation that will be called with the accumulator and the current element and should return the new accumulator
+ * @return the final accumulator value
+ */
+template<typename Acc, typename F, typename TupleLike>
```
same as above, TupleLike could be a Concept

bigerl commented on src/rdf4cpp/rdf/datatypes/registry/LiteralDatatypeImpl.hpp:
```diff
@@ -13,20 +13,23 @@
+} // namespace detail
 
-public:
+template<ConstexprString type_iri_t>
```
Add documentation for Default

bigerl commented on src/rdf4cpp/rdf/datatypes/registry/LiteralDatatypeImpl.hpp:
```diff
@@ -13,20 +13,23 @@
-    using always_false = std::false_type;
-    template<typename T>
-    static constexpr bool always_false_v = always_false<T>::value;
+namespace capabilities {
```
Have you consider super class `Capabilities<ConstexprString type_iri_t>`? You could also make Default superclass to the others. Please share your opinion in this.

liss-h replied:
I thought about it but I don't see what value that adds, since typedefs are not "inherited" by default, so
you have to put in a using decl anyways (if that was your intention).

bigerl replied:
yes, I forgot about the thing with the using declarations. I agree.

bigerl commented on src/rdf4cpp/rdf/Literal.cpp:
```diff
@@ -153,5 +155,225 @@ Literal Literal::make(std::string_view lexical_form, const IRI &datatype, Node::
+        auto const other_lit_type = other.handle_.literal_backend().datatype_id;
+
+        if (this_lit_type == other_lit_type) {
+            return std::make_tuple(this->datatype().identifier(), this->value(), other.value());
```
This and subsequent calls to `this->datatype().identifier()` are expensive because a lookup is made in the backend. But I don't see any portable way right now how to handle it differently.

bigerl commented on src/rdf4cpp/rdf/Literal.cpp:
```diff
@@ -153,5 +155,225 @@ Literal Literal::make(std::string_view lexical_form, const IRI &datatype, Node::
+            return std::make_tuple(this->datatype().identifier(), this->value(), other.value());
+        } else {
+            auto const lhs_datatype = this->datatype().identifier();
+            auto const rhs_datatype = other.datatype().identifier();
```
Sale here, use {this,other}_lit_type ti retrieve the identifier IRI

bigerl commented on src/rdf4cpp/rdf/Literal.cpp:
```diff
@@ -153,5 +155,225 @@ Literal Literal::make(std::string_view lexical_form, const IRI &datatype, Node::
+
+    auto const &[res_identifier, lhs_val, rhs_val] = *after_conversion;
+
+    auto const result_entry = datatypes::registry::DatatypeRegistry::get_entry(res_identifier).value().get();
```
Please double check that during every operation the *BackendView for each operand, and the type of each operand and, - if different - the type of the result is  retrieved at most once.

bigerl commented on src/rdf4cpp/rdf/Literal.cpp:
```diff
@@ -153,5 +155,225 @@ Literal Literal::make(std::string_view lexical_form, const IRI &datatype, Node::
+
+    assert(this->handle_.is_literal() && other.handle_.is_literal());
+
+    auto const after_conversion = [&]() -> std::optional<std::tuple<std::string_view, std::any, std::any>> {
```
That might be something for a seperate TODO: It would be better to use here the the datatype_id directly. 

Currently, the datatype_id is fixed only from some types but not in general. We would need to enforce a fixed datatype_id for all types that support non-trivial operatotions.

bigerl replied:
How about a sperate for AfterConversion? From the types alone it's not directly clear what the positions stand for.

bigerl commented on tests/nodes/tests_Literal_ops.cpp:
```diff
@@ -0,0 +1,429 @@
+    GENERATE_BINOP_TESTCASE(Integer, 1, +, Integer , 2, Integer, 3);
+    GENERATE_BINOP_TESTCASE(Integer, 12, -, Integer, 1, Integer, 11);
+    GENERATE_BINOP_TESTCASE(Integer, 3, *, Integer, 6, Integer, 18);
+    GENERATE_BINOP_TESTCASE(Integer, 12, /, Integer, 4, Integer, 3); // TODO: use idiv or div? currently using idiv
```
it must be div. idiv is not used. Instead: 

> but xsd:decimal if both operands are xsd:integer

My understanding would be to promote to `xsd:decimal` and do div there.

bigerl reviewed (changes requested):
thx for the splendid work. 
- I think the main point that needs to be adressed is minimizing the calls into NodeStorage. 
- Also, div for xsd:integer with  promotion to xsd:decimal, might need some more work. I have not yet double checked the actual behaviour.

bigerl commented on src/rdf4cpp/rdf/Literal.cpp:
```diff
@@ -153,5 +155,225 @@ Literal Literal::make(std::string_view lexical_form, const IRI &datatype, Node::
+
+    auto const result_entry = datatypes::registry::DatatypeRegistry::get_entry(res_identifier).value().get();
+
+    if (!result_entry.numeric_ops.has_value()) {
```
move up because we do not expect type hierarchies where we have non-numeric types which have might be cast to numeric types.

Clueliss committed:
added concept for tuple like types

Clueliss committed:
added static_assert to force preserving numericity in conversions

Clueliss committed:
added support for numeric operator result type specification
to allow div of integers to return decimal
and other pr review related improvements

Clueliss committed:
replaced lambdas with consteval functions in tuple concepts
to support order compilers

Clueliss committed:
added more explaination to static_assert for numericity

liss-h commented on src/rdf4cpp/rdf/datatypes/registry/DatatypeConversion.hpp:
```diff
@@ -0,0 +1,355 @@
+        // This is because the numeric operations assume that if any of the operands is not numeric
+        // the common type and the result type will also not be numeric to allow for an early return
+        // without doing the actual conversion.
+        static_assert(!NumericLiteralDatatype<Type> || NumericLiteralDatatype<typename next::converted>,
```
Technically this assertion is not strong enough because it means even if the result would be comparable
if one of the source types is not we do an early (null) return in the literal operations, which is potentially unexpected behaviour.

Should I change the one-sided implication to a two sided one? (So from -> to <->)
I mean it is technically not required for it to be two sided but maybe the behaviour would be more consistent(?).

Please share your thoughts.

bigerl replied:
I would say better be too strict than having undefined behavior.

bigerl commented on src/rdf4cpp/rdf/Literal.hpp:
```diff
@@ -40,12 +40,17 @@ class Literal : public Node {
      *
-     * @param logic_table the logic table for the binary operation accessed via logic_table[static_cast<size_t>(this->get_ebv_impl())][static_cast<size_t>(other.get_ebv_impl())]
+     * @param logic_table the logic table for the binary operation. It is accessed via
+     *                      logic_table[static_cast<size_t>(this->get_ebv_impl())][static_cast<size_t>(other.get_ebv_impl())].
```
any technical reason for the indentation?

liss-h replied:
nope, that was just an accident

bigerl commented on src/rdf4cpp/rdf/datatypes/registry/DatatypeRegistry.hpp:
```diff
@@ -29,25 +29,30 @@ class DatatypeRegistry {
 
-    using unop_fptr_t = std::any (*)(std::any const &);
-    using binop_fptr_t = std::any (*)(std::any const &, std::any const &);
+    struct NumericOpResult {
+        std::string_view result_type_iri;
+        std::any result_value;
+    };
```
Good idea to encapsulate both together.

bigerl commented on src/rdf4cpp/rdf/Literal.cpp:
```diff
@@ -157,78 +161,116 @@ Literal Literal::make(std::string_view lexical_form, const IRI &datatype, Node::
     }
 
-    auto op_res = op_select(*entry.numeric_ops)(this->value());
+    DatatypeRegistry::NumericOpResult const op_res = op_select(*entry->get().numeric_ops)(this->value());
```
`std::optional<std::reference_wrapper<X>>` is very verbose. Why not a raw pointer? Especially considering that the optional is not checked anyway.

liss-h replied:
Yes, you are right. I was thinking in rust not c++ there 😅 (in rust Option<&T> is the idiomatic thing)

bigerl commented on src/rdf4cpp/rdf/datatypes/registry/Util.hpp:
```diff
@@ -43,10 +91,10 @@ constexpr Acc tuple_type_fold_impl(Acc init, F &&f, std::index_sequence<Ixs...>)
-constexpr auto tuple_map(TupleLike &&tup, F &&f) {
-    constexpr size_t tup_size = std::tuple_size_v<std::remove_cvref_t<TupleLike>>;
-    return tuple_util_detail::tuple_map_impl(std::forward<TupleLike>(tup), std::forward<F>(f), std::make_index_sequence<tup_size>{});
+template<typename F, typename Tuple> requires AccessibleTuple<std::remove_cvref_t<Tuple>>
```
`std::remove_csref_t` would still leave e.g. volatile. I would always use `std::decay_t`. Same below

liss-h replied:
I don't think that is correct
see: https://en.cppreference.com/w/cpp/types/remove_cvref

bigerl replied:
Right, v is for volatile. After reading through the docs again, I would prefer `std::remove_csref_t` over `std:decay_t`. Easier to forsee what we will get out of applying it.

bigerl commented on src/rdf4cpp/rdf/Literal.cpp:
```diff
@@ -104,8 +108,8 @@ std::partial_ordering Literal::operator<=>(const Literal &other) const {
 bool Literal::operator==(const Literal &other) const {
-    if (this->null() && other.null()) {
-        return true;
+    if (this->null() || other.null()) {
```
good idea.

bigerl reviewed (changes requested):
I would say it is almost ready. Only some small things left to address. Please run format and optimize when everything else is done.

Clueliss committed:
replaced optional<reference_wrapper<DatatypeEntry const>> by DatatypeEntry const *

Clueliss committed:
stricter static assert to prevent unexpected behaviour

Clueliss committed:
added small optimization to allow reusing the already found entries for conversion

Clueliss committed:
fixed comment identation

Clueliss committed:
ran formatter

bigerl committed:
minor typos and missing doc part

bigerl committed:
add nodiscard to get_ebv_impl

bigerl reviewed (approved):
time to merge
